rust/abacus-base/src/contract_sync/mod.rs

```diff
@@ -47,3 +53,104 @@ impl<I> ContractSync<I> {
+    /// returns an inclusive range such as (0,50), (51,100), ...). This
+    /// will automatically rate limit based on how far we are from the
+    /// highest block we can scrape according to
+    /// `get_finalized_block_number`.
+    ///
+    /// In reality this will often return a from value that overlaps with the
+    /// previous range to help ensure that we scrape everything even if the
+    /// provider failed to respond in full previously.
+    ///
+    /// This assumes the caller will call next_range again automatically on Err,
+    /// but it returns the error to allow for tailored logging or different end
+    /// cases.
+    pub async fn next_range(&mut self) -> Result<(u32, u32)> {
+        self.rate_limit().await?;
+        let to = u32::min(self.tip, self.from + self.chunk_size);
+        let from = to.saturating_sub(self.chunk_size);
+        self.from = to + 1;
+        Ok((from, to))
+    }
+
+    /// If there was an issue when a range of data was fetched, this rolls back
+    /// so the next range fetched will be from `start_from`. Note that it is a
+    /// no-op if a later block value is specified.
+    pub fn backtrack(&mut self, start_from: u32) {
+        self.from = u32::min(start_from, self.from);
+    }
+
+    /// Wait based on how close we are to the tip and update the tip,
+    /// i.e. the highest block we may scrape.
+    async fn rate_limit(&mut self) -> Result<()> {
+        if self.from + self.chunk_size < self.tip {
+            // If doing the full chunk wouldn't exceed the already known tip,
+            // we don't necessarily need to fetch the new tip. Sleep a tiny bit
+            // so that we can catch up to the tip relatively quickly.
+            sleep(Duration::from_secs(1)).await;
+            Ok(())
+        } else {
+            // We are close to the tip.
+            if (Instant::now() - self.last_tip_update) < Duration::from_secs(30) {
+                // Sleep a little longer because we have caught up.
+                sleep(Duration::from_secs(10)).await;
+            } else {
+                // We are probably not caught up yet. This would happen if we
+                // started really far behind so now it is very likely the tip
+                // has moved a significant distance. We don't want to wait in
+                // this case any more than we normally would.
+                sleep(Duration::from_secs(1)).await;
+            }
+
+            match self.indexer.get_finalized_block_number().await {
+                Ok(tip) => {
+                    // we retrieved a new tip value, go ahead and update.
+                    self.last_tip_update = Instant::now();
+                    self.tip = tip;
+                    Ok(())
+                }
+                Err(e) => {
+                    // we are failing to make a basic query, we should wait before retrying.
+                    sleep(Duration::from_secs(10)).await;
```

Comment:
Benefit is backing off on the RPC provider if they are giving us errors for what should be a trivial request. The cause of such an error could be a few different things, but best practice is to not rapidly spam the same query if it is returning an error to give a chance for them to recover and avoid a retry-storm. Since this is a query that, in theory, should never fail, it seems reasonable to assume that an error here is indicating an issue on their end (or rate limiting).

Reply:
Ok, happy to keep it with the sleep